kardinal-cli/deployment/gateway.go

```diff
@@ -239,76 +223,59 @@ func isPodReady(pod *corev1.Pod) bool {
-	}
-	selector := strings.Join(labelSelectors, ",")
-
-	pods, err := client.CoreV1().Pods(namespace).List(context.Background(), metav1.ListOptions{LabelSelector: selector})
-	if err != nil {
-		return "", fmt.Errorf("error listing pods: %v", err)
-	}
-
-	if len(pods.Items) == 0 {
-		return "", fmt.Errorf("no pods found for service %s", service)
-	}
-
-	podName := pods.Items[0].Name
-	return podName, nil
-}
-
-// TODO move to the kubernetes package
-func checkGatewayEnvoyFilter(client *kubernetes.Clientset, host string) error {
-	for retry := 0; retry < maxRetries; retry++ {
-		envoyFilterRaw, err := client.RESTClient().
-			Get().
-			AbsPath("/apis/networking.istio.io/v1alpha3/namespaces/istio-system/envoyfilters/kardinal-gateway-tracing").
-			Do(context.Background()).
-			Raw()
+	var port int32
+	var namespace string
+	if accessEntry.Type == "ingress" {
+		ingress, err := client.NetworkingV1().Ingresses(accessEntry.Namespace).Get(context.Background(), accessEntry.Service, metav1.GetOptions{})
 		if err != nil {
-			logrus.Printf("Error getting Envoy filter (attempt %d/%d): %v", retry+1, maxRetries, err)
-			time.Sleep(retryInterval)
-			continue
+			return "", -1, "", fmt.Errorf("error getting ingress: %v", err)
 		}
 
-		var envoyFilter map[string]interface{}
-		err = json.Unmarshal(envoyFilterRaw, &envoyFilter)
+		ingressClassName := "nginx"
+		if ingress.Spec.IngressClassName != nil {
+			ingressClassName = *ingress.Spec.IngressClassName
+		}
+		ingressClass, err := client.NetworkingV1().IngressClasses().Get(context.Background(), ingressClassName, metav1.GetOptions{})
 		if err != nil {
-			logrus.Printf("Error unmarshaling Envoy filter (attempt %d/%d): %v", retry+1, maxRetries, err)
-			time.Sleep(retryInterval)
-			continue
+			return "", -1, "", fmt.Errorf("error getting ingress class: %v", err)
 		}
 
-		luaCode, ok := envoyFilter["spec"].(map[string]interface{})["configPatches"].([]interface{})[0].(map[string]interface{})["patch"].(map[string]interface{})["value"].(map[string]interface{})["typed_config"].(map[string]interface{})["inlineCode"].(string)
-		if !ok {
-			logrus.Printf("Error getting Lua code from Envoy filter (attempt %d/%d)", retry+1, maxRetries)
-			time.Sleep(retryInterval)
-			continue
+		for key, value := range ingressClass.Labels {
+			labelSelectors = append(labelSelectors, fmt.Sprintf("%s=%s", key, value))
 		}
+		namespace = ingressClass.Labels["app.kubernetes.io/instance"]
+		port = 80
 
-		if !strings.Contains(luaCode, host) {
-			logrus.Printf("Envoy filter 'kardinal-gateway-tracing' does not contain the expected host string: %s (attempt %d/%d)", host, retry+1, maxRetries)
-			time.Sleep(retryInterval)
-			continue
+	} else if accessEntry.Type == "gateway" {
+		gw, err := gwClient.GatewayV1().Gateways(accessEntry.Namespace).Get(context.Background(), accessEntry.Service, metav1.GetOptions{})
+		if err != nil {
+			return "", 0, "", fmt.Errorf("error getting gateway: %v", err)
```

Comment:
why in some error returns the port is 0 and 1 in other returns?

Reply:
Just a typo. But it should not matter since we are returning an error.

PS: I hate the lack of algebraic datatypes in golang